Desktop_Interface/uartstyledecoder.cpp

```diff
@@ -2,41 +2,43 @@
 #include <QDebug>
 #include <cassert>
 
-uartStyleDecoder::uartStyleDecoder(QObject *parent_in) : QObject(parent_in)
+uartStyleDecoder::uartStyleDecoder(QObject *parent_in)
+	: QObject(parent_in)
+	, parent{static_cast<isoBuffer*>(parent_in)}
+	, m_serialBuffer{SERIAL_BUFFER_LENGTH}
 {
-    parent = (isoBuffer *) parent_in;
 
 	// Begin decoding SAMPLE_DELAY seconds in the past.
 	serialPtr_bit = (int)(parent->m_back * 8 - SERIAL_DELAY * parent->m_sampleRate_bit + parent->m_bufferLen * 8) % (parent->m_bufferLen*8);
 
-    updateTimer = new QTimer();
-    updateTimer->setTimerType(Qt::PreciseTimer);
-    updateTimer->start(CONSOLE_UPDATE_TIMER_PERIOD);
-    connect(updateTimer, SIGNAL(timeout()), this, SLOT(updateConsole()));
+    m_updateTimer.setTimerType(Qt::PreciseTimer);
+    m_updateTimer.start(CONSOLE_UPDATE_TIMER_PERIOD);
+    connect(&m_updateTimer, &QTimer::timeout, this, &uartStyleDecoder::updateConsole);
 
-    serialBuffer = new isoBufferBuffer(SERIAL_BUFFER_LENGTH);
-
-    if(parent->m_channel == 1) console = parent->m_console1;
-    else if(parent->m_channel == 2) console = parent->m_console2;
-    else qFatal("Nonexistant console requested in uartStyleDecoder::serialDecode");
+    if (parent->m_channel == 1)
+		console = parent->m_console1;
+    else if (parent->m_channel == 2)
+		console = parent->m_console2;
+    else
+		qFatal("Nonexistant console requested in uartStyleDecoder::serialDecode");
 }
 
 uartStyleDecoder::~uartStyleDecoder()
 {
-    std::lock_guard<std::mutex> lock(mutex);
-	delete updateTimer;
-	delete serialBuffer;
 }
 
-void uartStyleDecoder::updateConsole(){
+void uartStyleDecoder::updateConsole()
+{
+	if (!newUartSymbol)
+		return;
+
     std::lock_guard<std::mutex> lock(mutex);
-	if(!newUartSymbol) return;
```

Comment:
Oh man, I cringed so hard reading the original code here.  Who locks a mutex _before_ checking a bool!?

Reply:
hahaha I have close to zero experience working with parallelism and synchronization, so I was really scared I would break something when i wrote this diff. I'm really happy I did the right thing!

Reply:
The lock_guard goes out of scope as soon as the function returns - you had nothing to worry about!
Now, if I'd used the `lock` method of std::mutex this would be a very different story (and you'd have probably fixed a bug).